### PR TITLE
*: use hakari to manage a workspace-hack crate

### PR DESCRIPTION
**Pull request by benesch**
@danhhz turns out I nerdsniped myself into doing this. 🙈 

----

workspace-hack crates speed up builds when working across a workspace with many crates by ensuring that all crates use the same features for all transitive dependencies. For details, see the excellent hakari documentation on the subject. [0]

This commit uses hakari to generate such a workspace-hack crate, and teaches CI to enforce that it is kept up to date.

[0]: https://docs.rs/cargo-hakari/latest/cargo_hakari/about/index.html

### Motivation

* Make local development a bit faster.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: ci/test/lint-fast.sh [hunk 1]
```diff
@@ -18,5 +18,7 @@ set -euo pipefail
 ci_try bin/lint
 ci_try cargo --locked fmt -- --check
 ci_try cargo --locked deny check licenses bans sources
+ci_try cargo hakari generate --diff
```
Comment (guswynn) [inline ci/test/lint-fast.sh:21]: <3

File: src/workspace-hack/build.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+fn main() {}
```
Comment (guswynn) [inline src/workspace-hack/build.rs:1]: why did you add a buildscript?

**Review by guswynn — APPROVED** [review]
Extremely dank.

did this not regress `cargo test` in `mz_persist_client` much?

**Review by danhhz — COMMENTED** [review]
persist-client currently has 501 transitive deps. looks like workspace-hack has 684. not ideal, but I'm willing to give it a shot! if it turns out to be a significant regression for persist iteration, maybe we could fall back to only opting in the leaf binary crates or something?

**Comment by benesch** [comment]
Rain pointed me at https://docs.rs/cargo-hakari/latest/cargo_hakari/config/index.html#final-excludes, which would allow us to exclude packages from the graph, like `mz-persist-client`. I'm kind of inclined to start by _excluding_ persist, and then y'all could choose to opt in if you want to choose the opposite tradeoff (faster environmentd compiles, but slower start-of-day persist compiles). WDYT, @danhhz?

**Comment by danhhz** [comment]
I only scanned that link, so possible I'm misunderstanding something, but that seems like it would be the world of all worlds for persist? I.e. we'd still depend on mz_ore, which would depend on the hack crate, so we'd still pull in all the transitive deps, but we wouldn't be included in the feature union calculation, so we'd still (potentially) have to recompile everything when switching between persist and bin/environmentd

**Comment by benesch** [comment]
Oh, right, good point. 😬

File: src/workspace-hack/build.rs (hunk 2 above)
Comment (benesch) [reply src/workspace-hack/build.rs:12]: Added a comment! It's required for Cargo to honor build dependencies.

**Comment by benesch** [comment]
Ok, @danhhz, going to merge as is. Let me know how the lived experience pans out. Happy to opt out `mz-ore` and `mz-persist-client`, or revert entirely, or something else.

**Comment by danhhz** [comment]
Will do! Quick unscientific test for `cargo build -p mz-persist-client --lib` with a clean target dir:

- with this change: 1m 55s
- before this change: 1m 35s

Certainly noticeable but only relevant when pulling past dep changes (plus the absolute numbers are small).
